src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/PowCoinviewRule.cs

```diff
@@ -125,7 +125,7 @@ public override Task RunAsync(RuleContext context)
             {
                 this.CheckBlockReward(context, fees, index.Height, block);
 
-                bool passed = checkInputs.All(c => c.GetAwaiter().GetResult());
+                bool passed = (await Task.WhenAll(checkInputs)).All(b => b);
```

Comment:
As above, you won't configure await with false

More importantly, however, it seems to me that you are losing a performance here. If I read the original code correctly,It will iterate through the tasks as long as they return true one by one, which means that is as 1 of the inputs is incorrect, the evaluation stops waiting for other tasks and we will return from the method

With your new implementation, this does not seem to hold as we will always wait for tasks to be completed before we iterate over their results.

Please correct me if I'm wrong because my linq–fu is weak.

I think we could very much improve the code even compared to the original version if we introduced the cancellation token and prevent the tasks to even start evaluation should the cancellation token be cancelled before they do so. This could be out of the scope of the PR, in which case please create an issue for that, however here we should definitely not make the things worse in this PR, so please replicate the original performance. Of course only if I read that correctly, if not please explain why it is the case because I would like to know.

Reply:
I don't seem to be able to find out definitively either way but yes it would appear that we would be waiting for all to finish before checking All whereas the original seems like it could shortcut. But then it comes down to whether the parallelism of the Task.When all outweighs that (perhaps in most cases all the tasks pass so awaiting all would be better?)

I can either return it to the original or perhaps

```
            foreach (Task<bool> checkInput in checkInputs)
            {
                if (await checkInput.ConfigureAwait(false) == false)
                {
                    ConsensusErrors.BadTransactionScriptError.Throw();
                }
            }
```
What do you think?

Reply:
Yes the for each is a simple solution, please do that and later we can improve using the cancellation token.

Please preserve the exit log, however:
```
 this.Logger.LogTrace("(-)[BAD_TX_SCRIPT]");
```

It is not better to wait for all every time because you need to think about the worst-case. So we want to quit checking as soon as possible to prevent denial of service attacks. This will be fully mitigated only when we later implement that cancellation token, but even here with just the for each instead of wait all we are at least no worse than original code.

Reply:
Yes of  course re the log line, I was just kinda psuedoing it

Reply:
Issue raised for cancellation token:  https://github.com/stratisproject/StratisBitcoinFullNode/issues/1446

Reply:
tx